Convert `AnnotationHeader` and its allies to preact

lyzadanger opened the pull request:
Marking WIP in case we want to break this up smaller.

This large PR migrates `AnnotationHeader` to preact and takes care of a number of related needs.

The former `AnnotationHeader` angular component packed a lot of stuff into a single component. As such, the new `AnnotationHeader` preact component makes use of three new sub-components. The first of these, `AnnotationUser`, is already extant. This PR adds two new components: `AnnotationShareInfo`, which renders info about an annotation's group and privacy status, and `AnnotationDocumentInfo` which renders metadata about an annotation's associated document (visible on the single-annotation and stream variants of annotation cards).

Those two subcomponents could be broken out into separate PRs.

Primarily, this migration is intended to preserve existing functionality, but there are a couple of changes to call out:

* The most visible change is that the icon used for annotations in open groups uses the "globe/public" icon also used in the publish control. This is more consistent and I think it gives a user more understanding of the visibility of a given annotation.
* This PR fixes a small glitch in which the timestamp component shows while the annotation is being edited (the intention is that it should not show during editing).
* The group, public (globe) and private icons are now SVG. We don't have an SVG icon for the "highlight" icon, so that is not replaced here.
* Small inconsistencies with spacing and rhythm are fixed.

## Before and After

### Sidebar Annotation Cards

#### Annotation in Public Group

This is an annotation in the Public group that is set to "Only Me" and is a highlight (that shows all possible icons). Note use of public/globe icon in new variant.

**BEFORE**

<img width="430" alt="sidebar-before-public" src="https://user-images.githubusercontent.com/439947/60453253-ad1f0300-9bfe-11e9-9291-ac4d75db892f.png">

**AFTER**

<img width="423" alt="sidebar-after-public" src="https://user-images.githubusercontent.com/439947/60453269-b90ac500-9bfe-11e9-88fd-30f752bc6377.png">

#### Annotation in Restricted or Private Group

This is an annotation in a restricted group (also "Only me" and a highlight to show icons for those). Private-group annotations look similar. Note new groups icon (matches publish control).

**BEFORE**

<img width="428" alt="sidebar-before-restricted" src="https://user-images.githubusercontent.com/439947/60453324-ddff3800-9bfe-11e9-9af3-d508a0511d91.png">

**AFTER**

<img width="426" alt="sidebar-after-restricted" src="https://user-images.githubusercontent.com/439947/60453331-e22b5580-9bfe-11e9-83c6-e2bc63100981.png">


#### Annotation in Third-party Group

This is an annotation in a third-party group. No sharing/group info is rendered. No changes here:

**BEFORE**

<img width="425" alt="sidebar-before-third-party" src="https://user-images.githubusercontent.com/439947/60453428-1ef74c80-9bff-11e9-8b91-0b2e495089e0.png">

**AFTER**

<img width="426" alt="sidebar-after-third-party" src="https://user-images.githubusercontent.com/439947/60453432-228ad380-9bff-11e9-8db8-d3c820df91cd.png">

#### "Only Me" Annotation in Third-Party Group

No group name displays, but "Only Me" text is rendered. It's subtle, but the new variant makes the spacing consistent:

**BEFORE**

<img width="428" alt="sidebar-before-third-party-only-me" src="https://user-images.githubusercontent.com/439947/60453476-451cec80-9bff-11e9-8ad9-5fd7ee05e5a9.png">

**AFTER**

<img width="423" alt="sidebar-after-third-party-only-me" src="https://user-images.githubusercontent.com/439947/60453483-49e1a080-9bff-11e9-8afe-52a5bc7cec68.png">

### Single Annotation or Stream Annotation Cards

These annotation cards show document metadata, and don't always have group information to render.

#### Annotation in Public Group

This annotation is the same annotation as the publicly-posted annotation above ("Annotation in Public Group"):

**BEFORE**

<img width="692" alt="single-before-public-only-me" src="https://user-images.githubusercontent.com/439947/60453585-84e3d400-9bff-11e9-94dc-f276f0454c6a.png">

**AFTER**

<img width="687" alt="single-after-public-only-me" src="https://user-images.githubusercontent.com/439947/60453600-8ad9b500-9bff-11e9-8a29-33c50d43d41f.png">

#### Annotation in Restricted Group

This is the same annotation as in the sidebar's "Annotation in Restricted or Private Group":

**BEFORE**

<img width="695" alt="single-before-restricted" src="https://user-images.githubusercontent.com/439947/60453645-b8bef980-9bff-11e9-8a83-99e4b1585bb1.png">

**AFTER**

<img width="691" alt="single-after-restricted" src="https://user-images.githubusercontent.com/439947/60453686-cd9b8d00-9bff-11e9-9def-b3b93da03f0c.png">

#### No Group Information Available

I don't want to make this even longer by explaining why this is the case (I can do that elsewhere), but in cases where the annotation is in a group (other than Public) that the current user can see but is not a member of, the group won't be available in single-annotation or stream views. Here is an example of such an annotation. This also shows the rendering of annotation document domain information, which is rendered when available. Observant folks will note that I ticked the domain font color down a notch as I felt it was hard to read before.

**BEFORE**

<img width="691" alt="single-before-no-group-only-me-domain" src="https://user-images.githubusercontent.com/439947/60453768-0cc9de00-9c00-11e9-969d-71a205f65cc7.png">

**AFTER**

<img width="689" alt="single-after-no-group-only-me-domain" src="https://user-images.githubusercontent.com/439947/60453787-17847300-9c00-11e9-900b-39c6a0acb16b.png">

robertknight commented on src/sidebar/components/test/annotation-header-test.js:
```diff
@@ -1,273 +1,155 @@
+    );
+  };
+
+  beforeEach(() => {
```
I can't see a reason why this mocking should be necessary. I deleted the `beforeEach` and `afterEach` blocks here and ran the tests and they passed. Can you try doing the same? If you get a failure, can you then try running `yarn --check-files` in case there is an issue with dependency versions. If that still fails, please let me know and we can try to figure out what is different between your environment and mine.

robertknight replied:
Ping

lyzadanger commented on src/sidebar/components/annotation-document-info.js:
```diff
@@ -0,0 +1,46 @@
+ * if a link is available.
+ */
+function AnnotationDocumentInfo({ annotation }) {
+  const docMetaMemoFn = memoize(annotationMetadata.domainAndTitle);
```
Is the general plan that we'd do away with the `memoize` util after migration, or does it have use for non-UI layers of the app indefinitely?

lyzadanger replied:
Regarding the use of `memoize` here, I was porting the Angular variant a little thoughtlessly. I'd imagine there must have been some sort of re-render issue with the document info (this hunch is made stronger by this line in the old template https://github.com/hypothesis/client/blob/d7a959c1ecdf9558a0839b9b3f355336b976a25e/src/sidebar/templates/annotation-header.html#L22 which suggests that someone was trying to squelch something). 

Unless there is some reason that document metadata would be flailing-ly updating on `annotation` objects without a meaningful change to the annotation itself, it doesn't seem like we should _need_ to memo-ize this here.

robertknight replied:
> Unless there is some reason that document metadata would be flailing-ly updating on annotation objects without a meaningful change to the annotation itself, it doesn't seem like we should need to memo-ize this here.

It is true that memoization in React/Preact is only an optimization, rather than being a requirement in some cases with Angular JS to cause its change detection loop to terminate. From a quick skim of the functions in `annotation-metadata.js` it looks like they do enough work that it would probably still be worth doing using `useMemo`. That's not essential to merge this though.

Something to be aware of is that React/Preact don't prevent skip re-rendering a component by default if the inputs haven't changed, they just avoid updating parts of the DOM where the output hasn't changed since the last render. I haven't done any profiling yet but I'm sure we'll  end up making use of [some of the methods available](https://reactjs.org/docs/optimizing-performance.html#avoid-reconciliation) to skip-rerendering of certain parts of the UI tree if inputs haven't changed.

lyzadanger commented on src/sidebar/components/annotation-header.js:
```diff
@@ -1,102 +1,91 @@
+  return (
+    <header className="annotation-header">
+      <div className="annotation-header__row">
+        <AnnotationUser annotation={annotation}></AnnotationUser>
```
Good catch. My editor's JSX linter tends to do a good job of most of this, so sometimes I overlook little stylistic things like this.

lyzadanger commented on src/sidebar/index.js:
```diff
@@ -141,7 +141,14 @@ function startAngularApp(config) {
     .component('annotation', require('./components/annotation'))
-    .component('annotationHeader', require('./components/annotation-header'))
+    .component(
+      'annotationDocumentInfo',
```
Right! Good catch again.

lyzadanger commented on src/styles/sidebar/components/annotation-header.scss:
```diff
@@ -0,0 +1,25 @@
+.annotation-header {
+  @include pie-clearfix;
```
I would dance happily if so. I shall try it and test in a few browsers.

lyzadanger replied:
Firefox, with clearfix in place:

<img width="466" alt="expected" src="https://user-images.githubusercontent.com/439947/60530138-203f7c80-9cc6-11e9-8d08-8b621495a7f6.png">

Firefox, with clearfix removed:

<img width="466" alt="firefox-without-clearfix" src="https://user-images.githubusercontent.com/439947/60530154-29304e00-9cc6-11e9-86e1-6fad6f74f192.png">


YIKES. Guess I'll leave it. Glad I tested in multiple browsers.

lyzadanger commented on src/sidebar/components/annotation-document-info.js:
```diff
@@ -0,0 +1,46 @@
+  return (
+    <div className="annotation-document-info">
+      <div className="annotation-document-info__title">
+        on &quot;
```
Wow, good observation! There are so many permutations of what order the bits can appear in depending on state that getting the margins/CSS right is a PhD-level endeavour, I swear.

lyzadanger commented on src/sidebar/components/annotation-share-info.js:
```diff
@@ -0,0 +1,77 @@
+   * The `links.html` property is only returned by the API for first-party
+   * groups, so this is a roundabout way of testing for that.
+   */
+  const linkToGroup = hasGroup && group.links && group.links.html;
```
Thanks for the fresh perspective. That's a good point and I will amend that comment.

lyzadanger commented:
Changes since last review:

* Not memoizing anything in `AnnotationDocumentInfo` for now
* Another pass on CSS sanity: adding margin on the right of elements only now, as there is a massive number of combinations and ordering here; fixes the nit here https://github.com/hypothesis/client/pull/1219#discussion_r299210054
* Fixed failing tests
* A few other small things

robertknight commented on src/sidebar/components/annotation-header.js:
```diff
@@ -1,102 +1,91 @@
+              className="annotation-header__timestamp-link"
+              href={annotationLink}
+              timestamp={annotation.updated}
+            ></Timestamp>
```
This can be replaced with a self-closing tag. You might want to consider enabling the [react/self-closing-comp](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/self-closing-comp.md) rule in our ESLint config. There is support for auto-fixing existing issues using `eslint --fix`.

robertknight commented on src/sidebar/components/annotation-share-info.js:
```diff
@@ -0,0 +1,71 @@
+ * whether it is private to the current user (only me)
+ */
+function AnnotationShareInfo({ annotation, groups, isPrivate }) {
+  const group = groups.get(annotation.group);
```
I suggest this should use `store.getGroup` via the `useStore` helper rather than `groups.get(<id>)` which is a proxy for that function to ensure that the component re-renders if the group changes.

lyzadanger commented:
Assuming Travis CI ever completes, ever, this is ready for a re-review @robertknight

robertknight commented on src/sidebar/components/annotation-header.js:
```diff
@@ -1,102 +1,84 @@
+  return (
+    <header className="annotation-header">
+      <div className="annotation-header__row">
+        <AnnotationUser annotation={annotation} />
```
I found an issue here that the height of the top row reduces from 20px to 17px when editing an annotation. This happens because the `<div>`s in the row have a height of 20px, which comes from their computed `line-height` property. When the annotation is opened, the `<div>`s are removed and so the height of the row drops to that of the shortest item, the username link.

Its not a very explicit solution, but one way to resolve this is to wrap the `<AnnotationUser>` in a plain `<div>`.

lyzadanger replied:
Thank you for the clear explanation of this! Very helpful.

I'm squinting at `AnnotationUser` now and thinking it ain't great that it returns _either_ an `a` or a `div` element depending on context. I'm thinking it should always return a `div` but want to squint a little harder for a while.

robertknight commented on src/sidebar/components/annotation-header.js:
```diff
@@ -1,102 +1,84 @@
+          </a>
+        </div>
+        {!isEditing && annotation.updated && (
+          <div className="annotation-header__timestamp">
```
Since the row is using flex layout, any text/inline items are wrapped in an anonymous block container. You could therefore apply the `margin-left: auto` property of the `annotation-header__timestamp` div directly to the `annotation-header__timestamp-link` class.

However, see the caveat about the height of this row changing above when there are no divs in the line, only text/inline items.

robertknight commented on src/sidebar/components/test/annotation-header-test.js:
```diff
@@ -1,273 +1,155 @@
-        );
-        assert.equal(ctrl.htmlLink(), '');
-      });
+      assert.isOk(highlightIcon.exists());
```
This is OK, though personally I prefer to use `assert.{isTrue, isFalse}` rather than `assert.{isOk, notOk}` unless I explicitly intend to check for falseyness, as opposed to actual boolean values.

robertknight commented on src/sidebar/components/test/annotation-header-test.js:
```diff
@@ -1,273 +1,155 @@
-          }
-        );
-        assert.deepEqual(ctrl.documentMeta(), fakeDocumentMeta);
+    it('should not display the is-highlight icon if annotation is not a highlight', () => {
```
Nit: Would you mind adding new lines between tests for consistency with the rest of the file.

robertknight commented on src/styles/sidebar/components/annotation-document-info.scss:
```diff
@@ -0,0 +1,14 @@
+.annotation-document-info {
+  font-size: 13px;
```
I know that our typography stack is not well defined at the moment, but I think it would still be preferable to use the font-size variables that do exist from the "Typography" section in `variables.scss` rather than hard-coding them here. That will help us establish at least which things are supposed to be the same size and adjust them uniformly.

lyzadanger commented:
@robertknight Regarding the problematic `withServices` subcomponents vis-a-vis mocking, I can't reproduce the failures now! Bizarre. I'll keep an eye out.

lyzadanger commented:
@robertknight OK! Stuff has been addressed. I've updated `AnnotationUser` to always render a `div` (versus a bare `a` in some cases), which fixes the `line-height` problem you observed.
